File: packages/contracts/src/telemetry.ts
import { z } from 'zod';

export const PrometheusClientConfigSchema = z.object({
	baseUrl: z.string().url().optional(),
	bearerToken: z.string().optional(),
	timeout: z.string().optional(),
});

export const GateRunContextSchema = z.object({
	flagKey: z.string(),
	treatmentVariationId: z.string(),
	controlVariationId: z.string(),
	environmentKey: z.string().optional(),
});

export const GateVerdictResultSchema = z.enum(['pass', 'fail']);

export const GateEvaluationResultSchema = z.object({
	verdict: GateVerdictResultSchema,
	metricType: z.string(),
	observedDelta: z.number().optional(),
	treatmentValue: z.number().optional(),
	controlValue: z.number().optional(),
	threshold: z.number(),
	metadata: z.record(z.string(), z.unknown()),
});

export const PreflightCheckSchema = z.object({
	id: z.string(),
	status: z.enum(['pass', 'fail']),
	detail: z.string().optional(),
	observed: z.number().optional(),
	required: z.number().optional(),
});

export const PreflightReportSchema = z.object({
	status: z.enum(['pass', 'fail']),
	checks: z.array(PreflightCheckSchema),
	blockReason: z.string().optional(),
});

export const StageGateEvaluationSchema = z.object({
	verdict: GateVerdictResultSchema,
	results: z.array(GateEvaluationResultSchema),
});

export type PrometheusClientConfig = z.infer<typeof PrometheusClientConfigSchema>;
export type GateRunContext = z.infer<typeof GateRunContextSchema>;
export type GateVerdictResult = z.infer<typeof GateVerdictResultSchema>;
export type GateEvaluationResult = z.infer<typeof GateEvaluationResultSchema>;
export type PreflightCheck = z.infer<typeof PreflightCheckSchema>;
export type PreflightReport = z.infer<typeof PreflightReportSchema>;
export type StageGateEvaluation = z.infer<typeof StageGateEvaluationSchema>;

File: kotlin/modules/contracts/src/main/kotlin/com/ffpromo/contracts/Telemetry.kt
package com.ffpromo.contracts

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class PrometheusClientConfig(
    val baseUrl: String? = null,
    val bearerToken: String? = null,
    val timeout: String? = null,
)

@Serializable
data class GateRunContext(
    val flagKey: String,
    val treatmentVariationId: String,
    val controlVariationId: String,
    val environmentKey: String? = null,
)

@Serializable
enum class GateVerdictResult {
    @SerialName("pass")
    pass,

    @SerialName("fail")
    fail,
}

@Serializable
enum class PreflightCheckStatus {
    @SerialName("pass")
    pass,

    @SerialName("fail")
    fail,
}

@Serializable
data class GateEvaluationResult(
    val verdict: GateVerdictResult,
    val metricType: String,
    val observedDelta: Double? = null,
    val treatmentValue: Double? = null,
    val controlValue: Double? = null,
    val threshold: Double,
    val metadata: Map<String, JsonElement> = emptyMap(),
)

@Serializable
data class PreflightCheck(
    val id: String,
    val status: PreflightCheckStatus,
    val detail: String? = null,
    val observed: Double? = null,
    val required: Double? = null,
)

@Serializable
data class PreflightReport(
    val status: PreflightCheckStatus,
    val checks: List<PreflightCheck>,
    val blockReason: String? = null,
)

@Serializable
data class StageGateEvaluation(
    val verdict: GateVerdictResult,
    val results: List<GateEvaluationResult>,
)
